projects/src/transmitter.cc

```diff
@@ -0,0 +1,157 @@
+}
+
+void Transmitter::encode_message(msg_t msg) {
+    // For now, we send a simulation message but this needs to be fixed later on.
+    // TODO: Presend protobuf tag. See https://github.com/oliverlee/phobos/issues/181#issuecomment-301825244
+    m_bytes_written = 0;
+    if ((msg % sizeof(stkalign_t)) == 0) {
+        if ((reinterpret_cast<BicyclePoseMessage*>(msg) >= m_pose_message_buffer) &&
+            (reinterpret_cast<BicyclePoseMessage*>(msg) <= &m_pose_message_buffer[POSE_MESSAGE_POOL_SIZE])) {
+            SimulationMessage sim_msg = SimulationMessage_init_zero;
+            sim_msg.timestamp = 0;
+            sim_msg.pose = *reinterpret_cast<BicyclePoseMessage*>(msg);
+            sim_msg.has_pose = true;
+            free_pose_message(reinterpret_cast<BicyclePoseMessage*>(msg));
+            m_bytes_written = encode_packet(sim_msg);
+        } else if (
+                (reinterpret_cast<SimulationMessage*>(msg) >= m_simulation_message_buffer) &&
+                (reinterpret_cast<SimulationMessage*>(msg) <= &m_simulation_message_buffer[SIMULATION_MESSAGE_POOL_SIZE])) {
+            m_bytes_written = encode_packet(*reinterpret_cast<SimulationMessage*>(msg));
+            free_simulation_message(reinterpret_cast<SimulationMessage*>(msg));
+        } else {
+            chDbgAssert(false, "msg pointer does not originate from a valid memory pool");
+        }
+    } else {
+        chDbgAssert(false, "msg pointer is not stack aligned");
+    }
+}
+
+void Transmitter::transmit_packet() const {
+    // TODO: Change usbTransmit to usbStartTransmitI and encode during USB transmission?
+    //       This may result in a single USB transmission buffer containing more than one packet.
+    usbTransmit(SDU1.config->usbp, SDU1.config->bulk_in, m_packet_buffer.data(), m_bytes_written);
+}
+
+size_t Transmitter::encode_packet(const SimulationMessage& m) {
+   const size_t serialize_buffer_len = packet::serialize::encode_delimited(
+       m, m_serialize_buffer.data(), m_serialize_buffer.size());
+
+   const cobs::EncodeResult encode_result = cobs::encode(
+       m_serialize_buffer.data(), serialize_buffer_len, m_packet_buffer.data(), m_packet_buffer.size());
+
+   // Encoding only fails when the destination buffer is too small, this
+   // should not happen as long as we only encode SimulationMessage objects
+   // because we allocated m_packet_buffer based on its size.
+   chDbgAssert(encode_result.status == cobs::EncodeResult::Status::OK, "Expected encoding to succeed.");
+
+   return encode_result.produced;
+}
+
+void Transmitter::transmitter_thread_function(void* p) {
+    auto self = static_cast<Transmitter*>(p);
+
+    chRegSetThreadName("transmitter");
+    self->initialize();
+
+    while (!chThdShouldTerminateX()) {
+        msg_t msg = reinterpret_cast<msg_t>(nullptr);
+        chMBFetch(&self->m_message_mailbox, &msg, TIME_INFINITE);
+        self->encode_message(msg);
```

Comment:
`encode_message` and `transmit_message` are called after each other every time. Because they depend on some instance state, I'd consider merging them into 1 function. Even thought they are separate steps conceptually, they are always both performed in the same sequence.